### PR TITLE
Implement /track_growth

### PR DESCRIPTION
**Pull request by BabarKhanDev**

### CODE REVIEW

File: app/main.py [hunk 1]
```diff
@@ -113,8 +116,55 @@ async def upload_image(file: UploadFile = File(...)):
...
-@app.get("/track_growth")
-async def track_growth():
-    raise HTTPException(
-        status_code=status.HTTP_501_NOT_IMPLEMENTED
-    )
+@app.get("/track_growth/")
+async def track_growth(center: Tuple[float, float] = None, scan_range: float = 360, day_range: int = 1000):
+
+    if center is None:
+        center = (0, 0)
```
Comment (Laura7089) [inline app/main.py:123]: This is equivalent to making this a default in the function args

File: app/main.py [hunk 2]
```diff
@@ -113,8 +116,55 @@ async def upload_image(file: UploadFile = File(...)):
...
+            t2.latitude,
+            t2.longitude,
+            COUNT(*) AS count
+        FROM 
+            image_processing.processed_entry t1
+        JOIN 
+            image_processing.raw_entry t2 ON t1.image_uri = t2.image_uri
+        WHERE
+            (({center[0]} + t2.latitude)^2 + ({center[1]} + t2.longitude)^2) <= {scan_range}^2
+            AND t2.date BETWEEN CURRENT_DATE - {day_range} AND CURRENT_DATE
```
Comment (Laura7089) [inline app/main.py:149]: What is the extra `AND CURRENT_DATE` doing?

File: app/main.py [hunk 3]
```diff
@@ -113,8 +116,55 @@ async def upload_image(file: UploadFile = File(...)):
...
+            t1.plant_id, t2.date, t2.latitude, t2.longitude
+    ) AS subquery
+    GROUP BY 
+        plant_id;
+    """
+
+    def format_datum(datum):
+        return [PlantInstanceData(
+            date=date,
+            latitude=float(lat),
```
Comment (Laura7089) [inline app/main.py:160]: Why get the formatted values with `::text` (I believe that's what that does) if you're only gonna re-convert them here?

File: util/data_generators.py [hunk 4]
```diff
@@ -25,11 +25,16 @@ def growth_map(
 
 
 def generate_entry(conn, date: datetime, plant_id: str) -> None:
-    image_uri = uuid.uuid4().hex
+    # TODO in demo, put the location and range of an actual forest here
```
Comment (Laura7089) [inline util/data_generators.py:28]: Perhaps a way to pass these in so we aren't modifying code specifically for the demo?

File: app/main.py [hunk 5]
```diff
@@ -113,8 +116,55 @@ async def upload_image(file: UploadFile = File(...)):
         ])
 
 
-@app.get("/track_growth")
-async def track_growth():
-    raise HTTPException(
-        status_code=status.HTTP_501_NOT_IMPLEMENTED
-    )
+@app.get("/track_growth/")
```
[the remark this answers is not in the document]
Comment (BabarKhanDev) [reply app/main.py:119]: The docs made me think it was necessary, just checked and it's not. Removed :)

File: app/main.py (hunk 1 above)
Comment (BabarKhanDev) [reply app/main.py:123]: Ah my bad I thought tuples were mutable but they're not, fixed :)

File: app/main.py (hunk 2 above)
Comment (BabarKhanDev) [reply app/main.py:149]: It's checking for all entries between (current date - day range) and the current date

File: app/main.py [hunk 6]
```diff
@@ -113,8 +116,55 @@ async def upload_image(file: UploadFile = File(...)):
...
+
+    if center is None:
+        center = (0, 0)
+
+    SQL = f"""
+    SELECT 
+        plant_id, 
+        ARRAY_AGG(
+            ARRAY[
+                TO_CHAR(DATE_TRUNC('day', date), 'YYYY-MM-DD'),
```
[the remark this answers is not in the document]
Comment (BabarKhanDev) [reply app/main.py:130]: There's some delay between the scan and when they get uploaded to the db, I figured any higher precision would be inaccurate

[the remark this answers is not in the document]
Comment (Laura7089) [reply app/main.py:130]: You're implicitly dropping anything that's not Year/Month/Day when you string format this iiuc so it should have the same effect

File: app/main.py (hunk 3 above)
Comment (BabarKhanDev) [reply app/main.py:160]: all elements in a postgresql array need to be the same type (or so I think) so I pull them all out as strings and then convert them back to their proper types for the front end

File: app/main.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (BabarKhanDev) [reply app/main.py:130]: I misunderstood what you meant, fixed now :)

File: util/data_generators.py (hunk 4 above)
Comment (BabarKhanDev) [reply util/data_generators.py:28]: added :)

**Review by Laura7089 — APPROVED** [review]

**Pull request merged**